README
Diffuse
Version 0.4.6

Copyright (C) 2006-2011 Derrick Moser <derrick_moser@yahoo.com>

----

Diffuse is a graphical tool for merging and comparing text files.  Diffuse is
able to compare an arbitrary number of files side-by-side and gives users the
ability to manually adjust line matching and directly edit files.  Diffuse can
also retrieve revisions of files from Bazaar, CVS, Darcs, Git, Mercurial,
Monotone, RCS, Subversion, and SVK repositories for comparison and merging.

Some key features of Diffuse:
- ability to compare and merge an arbitrary number of files side-by-side (n-way merges)
- line matching can be manually corrected by the user
- ability to directly edit files
- syntax highlighting
- Bazaar, CVS, Darcs, Git, Mercurial, Monotone, RCS, Subversion, and SVK support
- Unicode support
- unlimited undo
- easy keyboard navigation

----

Requirements:

- Python >= 2.4
- PyGTK >= 2.10

----

Diffuse is implemented entirely in Python and should run on any platform with
Python and PyGTK.  If you need to manually install PyGTK, please be aware its
dependencies should be installed prior to installing PyGTK.

Diffuse can be run directly from an untared source distribution on any POSIX
system or installed with the instructions described in the next section.

The location of the personal preferences, state, and initialisation files have
changed in the 0.4.1 release.  Old settings may be migrated using the following
commands:

  $ mkdir -p ~/.config/diffuse
  $ mv ~/.diffuse/config ~/.config/diffuse/state
  $ mv ~/.diffuse/* ~/.config/diffuse
  $ rmdir ~/.diffuse

The rules for parsing files in ~/.diffuse changed in the 0.3.0 release.
Non-fatal errors may be reported when parsing old files.  These errors can be
fixed by removing the offending lines (or the entire file) from
~/.config/diffuse/diffuserc.

----

Installing on POSIX systems:

Run the 'install.py' script to install Diffuse.  The script accepts several
options for customising the installation behaviour.  Use the '--help' option
to obtain a full listing.

Diffuse can be removed by running the 'install.py' script with the '--remove'
option.

Specify the '--destdir=' and '--files-only' options for the 'install.py'
script when using it to build a platform specific package of Diffuse.

----

Installing on Windows:

The 'windows-installer' directory contains scripts for building an installable
package for Windows that includes all dependencies.

Diffuse can be packaged as a portable application by copying the installation
directory to a pen drive and creating a front end that sets the XDG_CONFIG_HOME
and XDG_DATA_DIR environment variables prior to launching Diffuse.  The
XDG_CONFIG_HOME and XDG_DATA_DIR environment variables indicate where Diffuse
should store persistent settings (eg. the path to a writable directory on the
pen drive).

----

Help Documentation:

Diffuse's help documentation is written in the DocBook format and can be easily
converted into other formats using XSLT stylesheets.  If the local help
documentation or its browser are unavailable, Diffuse will attempt to display
the on-line help documentation using a web browser.


ChangeLog
- added Pedro Albuquerque's Portuguese translation
- added Åke Engelbrektson's Swedish translation
- added Guillaume Hoffmann's Darcs support improvements
- add support for Git submodules
- added Akom Chotiphantawanon's Thai translation
- updated Python highlighting for Python 3 grammar
- added a preference and command line option to specify the version control system search order

Diffuse 0.4.8 - 2014-07-18
- updated use of gtk.SpinButton and gtk.Entry to avoid quirks seen on some platforms
- fixed a bug that prevented Diffuse from reviewing large git merge conflicts
- updated C/C++ syntax highlighting to recognise C11/C++11 keywords
- fixed a bug that prevented drag-and-drop of file paths with non-ASCII characters
- improved image quality of icons
- added Chi Ming and Wei-Lun Chao's Traditional Chinese translation

Diffuse 0.4.7 - 2013-05-13
- added Jindřich Šesták's Czech translation
- improved character editing to allow easy indenting and moving the cursor by whole words
- added Miś Uszatek's Polish translation
- improved auto-detection of utf_16 and utf_32
- added "New N-Way File Merge..." menu item
- added syntax highlighting for Erlang and OpenCL files

Diffuse 0.4.6 - 2011-11-02
- added support for Subversion 1.7
- added "Open Commit..." menu item
- "-c" option now works for all supported version control systems
- fixed a bug that caused the wrong revision to be shown when working on a branch in Mercurial
- Git support distinguishes between staged and unstaged files
- added syntax highlighting for R files

Diffuse 0.4.5 - 2011-07-13
- fixed a bug in CVS and Subversion support that prevented Diffuse from displaying some removed files
- added syntax highlighting for JSON files
- added menu items and keyboard shortcuts for "First Tab" and "Last Tab"
- added "--line" command line option
- fixed a bug that caused deleted files to be ignored when using the '-m' option
- fixed a bug that incorrectly encoded pasted text if utf_8 was not specified in the Region Settings preferences
- state information is now stored in ~/.local/share/diffuse
- Diffuse now uses a patience diff-based algorithm to align lines
- added command line option to specify a label to display instead of the file name
- added preference to display the right margin
- added Cristian Marchi's Italian translation
- fixed a bug that could cause "Save As..." to fail with some user specified encodings

Diffuse 0.4.4 - 2010-10-21
- Git support now recognises conflicts when re-applying the stash
- double clicking on text can now select full words with non-English characters
- search dialog is now automatically populated with the currently selected text
- fixed a bug that prevented opening files with non-ASCII characters in their path
- added Oleg Pakhtusov's Russian translation
- added Kang Bundo's Korean translation
- pane headers tooltips
- Shift-ScrollWheel can now be used to scroll horizontally

Diffuse 0.4.3 - 2010-04-15
- fixed a bug that prevented the "-m" option from opening a 3-way merge for Subversion and Bazaar conflicts

Diffuse 0.4.2 - 2010-04-13
- support for detached Git repositories
- better removal of unnecessary spacer lines
- added support for horizontal mouse scrolling
- renamed some resources to more user friendly names
- RCS support
- added Henri Menke's Spanish translation
- "#!" interpreter lines are now used to select proper highlighting rules
- syntax highlighting is now indicated by radio menu items

Diffuse 0.4.1 - 2009-10-12
- added Japanese translation
- added Liu Hao's simplified Chinese translation
- added a 'Dismiss All Edits' menu item
- personal configuration files are now stored in ~/.config/diffuse/ (the README file describes how to migrate old settings)
- localised manuals are now supported on Windows
- new command line option for specifying blank file comparison panes
- new preference to enable/disable line numbers
- Diffuse now quits if no viewers were created with the -m option
- MMB on a notebook tab now closes the tab
- RMB on a notebook tab creates a popup menu to set the current page
- added "Undo Close Tab" menu item
- replaced "Closing last tab quits Diffuse" preference with a "Warn me when closing a tab will quit Diffuse" preference
- added new menu items and buttons for copying lines between panes
- changed the default hotkeys for merging to reflect the direction text "moves"

Diffuse 0.4.0 - 2009-08-17
- added format menu with new items for changing case, sorting, and manipulating white space
- replaced 'Hide end of line characters' preference with 'Show white space characters'
- errors are now reported in a dialogue instead of printing to stderr
- optimised redraws when only the cursor position has changed
- input methods that use pre-editing can now be used when editing text
- dead keys can now be used when editing text
- updated Monotone support to use 'mtn automate inventory'
- Git support now handles files flagged as 'unmerged'
- graceful handling of and non-zero exit codes from 'git status'
- added version control section to the manual
- minor bug fixes

Diffuse 0.3.4 - 2009-07-03
- syntax highlighting for .plist, GLSL, SConscript, and SConstruct files
- status bar now explains how to navigate between modes
- added labels to indicate syntax highlighting rules, encoding, and format
- Subversion 1.6 support
- added Henri Menke's German translation
- added '--examplesdir=' and '--mandir=' options to install.py
- renamed the '--python-interpreter=' installer option to '--pythonbin='
- minor bug fixes

Diffuse 0.3.3 - 2009-04-13
- fixed a bug handling the backspace key with the cursor in the first column

Diffuse 0.3.2 - 2009-04-13
- POSIX installer with '--destdir=' and '--files-only' options for packagers
- vi-like keybindings for line mode
- '-m' option to open modified files in separate tabs
- 'Merge From Left Then Right' and 'Merge From Right Then Left' menu items
- drag-n-drop support
- preferences for behaviour of tabs
- files with edits now tagged with '*'
- auto indent

utils/makemanual.py
#!/usr/bin/env python2

# Copyright (C) 2010 Derrick Moser <derrick_moser@yahoo.com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the licence, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program.  You may also obtain a copy of the GNU General Public License
# from the Free Software Foundation by visiting their web site
# (http://www.fsf.org/) or by writing to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

# This program translates Diffuse's DocBook help into a manual page using the
# book2manual.xsl template.

import os
import subprocess

# fetch translations for English text hard coded in the stylesheets
translations = {}
f = open('translations.txt', 'rb')
for v in f.read().split('\n'):
    v = v.split(':')
    if len(v) == 3:
        lang = v[0]
        if not translations.has_key(lang):
            translations[lang] = []
        translations[lang].append(v[1:])
f.close()

# find all localised versions of the DocBook manual
d = '../src/usr/share/gnome/help/diffuse'
for lang in os.listdir(d):
    # skip over any .svn files
    if lang.startswith('.'):
        continue

    # convert regular DocBook manual into a form suitable for man pages
    cmd = [ 'xsltproc', 'book2manual.xsl', os.path.join(os.path.join(d, lang), 'diffuse.xml') ]
    proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    proc.stdin.close()
    proc.stderr.close()
    fd = proc.stdout
    s = fd.read()
    fd.close()
    if proc.wait() != 0:
        raise OSError('Could not run xsltproc')
 
    # save man page suitable DocBook file
    a = 'diffuse.xml'
    f = open(a, 'wb')
    f.write(s)
    f.close()

    # convert to man page
    cmd = [ 'xsltproc', '/usr/share/sgml/docbook/stylesheet/xsl/nwalsh/manpages/docbook.xsl', a ]
    proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    proc.stdin.close()
    proc.stderr.close()
    fd = proc.stdout
    s = fd.read()
    fd.close()
    if proc.wait() != 0:
        raise OSError('Could not run xsltproc')
    os.unlink(a)

    # read resulting man page
    f = open('diffuse.1', 'rb')
    s = f.read()
    f.close()
    os.unlink('diffuse.1')
    # remove comments
    s = '\n'.join([ c for c in s.split('\n') if not c.startswith('.\\"') ])
    # fix arrow
    s = s.replace('\xe2\x86\x92', '\\(->')
    # translate extra text
    for k, v in translations.get(lang, []):
        s = s.replace(k, v)

    # figure out where to save the new man page
    # make subdirectories if necessary
    dd = '../src/usr/share/man'
    if lang != 'C':
        dd = os.path.join(dd, lang)
    if not os.path.exists(dd):
        os.mkdir(dd)
    dd = os.path.join(dd, 'man1')
    if not os.path.exists(dd):
        os.mkdir(dd)

    # finally save the man page
    f = open(os.path.join(dd, 'diffuse.1'), 'wb')
    f.write(s)
    f.close()


utils/makemanual.sh
#!/bin/bash

# Copyright (C) 2009-2010 Derrick Moser <derrick_moser@yahoo.com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the licence, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program.  You may also obtain a copy of the GNU General Public License
# from the Free Software Foundation by visiting their web site
# (http://www.fsf.org/) or by writing to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

# This program translates Diffuse's DocBook help into a manual page using the
# book2manual.xsl template.

if [ "$1" == "" ]; then
    echo "Usage: $0 <path_to_docbook_xsl_home>"
    echo "  eg. $0 /usr/share/sgml/docbook/stylesheet/xsl/nwalsh"
    echo "  eg. $0 /usr/share/sgml/docbook/xsl-stylesheets--1.73.2"
    exit 1
fi

# create man page
xsltproc --nonet book2manual.xsl ../src/usr/share/gnome/help/diffuse/C/diffuse.xml | xsltproc --nonet "$1/manpages/docbook.xsl" -
# eliminate comment lines and replace arrow character with correct sequence
sed -i '/^\.\\"/d;s/\xe2\x86\x92/\\(->/g' diffuse.1

# create on-line manual
# eliminate ugly empty paragraphs
xsltproc "$1/html/docbook.xsl" ../src/usr/share/gnome/help/diffuse/C/diffuse.xml | sed ':a;N;$!ba;s/<p>[ \n]*<\/p>//g' > manual.html


windows-installer/build.py
# Copyright (C) 2006-2014 Derrick Moser <derrick_moser@yahoo.com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the licence, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program.  You may also obtain a copy of the GNU General Public License
# from the Free Software Foundation by visiting their web site
# (http://www.fsf.org/) or by writing to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

# This program builds a Windows installer for Diffuse.

import codecs
import glob
import os
import platform
import subprocess
import sys

VERSION='0.4.7'
PACKAGE='1'
PLATFORM='win' + ''.join([ c for c in platform.architecture()[0] if c.isdigit() ])
INSTALLER='diffuse-%s-%s.%s' % (VERSION, PACKAGE, PLATFORM)

# makes a directory without complaining if it already exists
def mkdir(s):
    if not os.path.isdir(s):
        os.mkdir(s)

# copies a file to 'dest'
def copyFile(src, dest, use_text_mode=False,enc=None):
    print 'copying "%s" to "%s"' % (src, dest)
    if use_text_mode:
        r, w = 'r', 'w'
    else:
        r, w = 'rb', 'wb'
    f = open(src, r)
    s = f.read()
    f.close()
    if enc is not None:
        s = codecs.encode(unicode(s, 'utf_8'), enc)
    f = open(dest, w)
    f.write(s)
    f.close()

# recursively copies a directory to 'dest'
def copyDir(src, dest):
    print 'copying "%s" to "%s"' % (src, dest)
    mkdir(dest)
    for f in os.listdir(src):
        s = os.path.join(src, f)
        d = os.path.join(dest, f)
        if os.path.isfile(s):
            copyFile(s, d)
        elif os.path.isdir(s):
            copyDir(s, d)

# helper to clean up the resulting HTML
def extract_tag(s, start, end):
    i = s.find(start)
    if i >= 0:
        pre = s[:i]
        i += len(start)
        j = s.find(end, i)
        if j >= 0:
            return pre, start, s[i:j], end, s[j+len(end):]

#
# Make sure we are in the correct directory.
#

path = os.path.dirname(sys.argv[0])
if path != '':
    os.chdir(path)

#
# Build EXE versions of the Diffuse Python script.
#

# make a temp directory
mkdir('temp')
# copy script into temp directory under two names
for p in 'temp\\diffuse.py', 'temp\\diffusew.pyw':
    copyFile('..\\src\\usr\\bin\\diffuse', p, True)

# build executable in 'dist' from diffuse.py and diffusew.pyw
args = [ sys.executable, 'setup.py', 'py2exe' ]
if os.spawnv(os.P_WAIT, args[0], args) != 0:
    raise OSError('Could not run setup.py')

# include Microsoft redistributables needed by Python 2.6 and above
for f in 'msvcm90.dll', 'msvcp90.dll', 'msvcr90.dll':
    copyFile(os.path.join(os.environ['SYSTEMROOT'], 'WinSxS\\x86_Microsoft.VC90.CRT_1fc8b3b9a1e18e3b_9.0.21022.8_x-ww_d08d0375\\' + f), 'dist\\' + f)
copyFile(os.path.join(os.environ['SYSTEMROOT'], 'WinSxS\\Manifests\\x86_Microsoft.VC90.CRT_1fc8b3b9a1e18e3b_9.0.21022.8_x-ww_d08d0375.manifest'), 'dist\\Microsoft.VC90.CRT.manifest')

# include GTK dependencies
gtk_dir = os.environ['GTK_BASEPATH']
copyDir(os.path.join(gtk_dir, 'etc'), 'dist\\etc')
copyDir(os.path.join(gtk_dir, 'lib'), 'dist\\lib')
mkdir('dist\\share')
copyDir(os.path.join(gtk_dir, 'share\\icons'), 'dist\\share\\icons')
copyDir(os.path.join(gtk_dir, 'share\\themes'), 'dist\\share\\themes')

#
# Add all support files.
#

# syntax highlighting support
mkdir('dist\\syntax')
for p in glob.glob('..\\src\\usr\\share\\diffuse\\syntax\\*.syntax'):
    copyFile(p, os.path.join('dist\\syntax', os.path.basename(p)), True)
copyFile('diffuserc', 'dist\\diffuserc')

# application icon
copyDir('..\\src\\usr\\share\\icons', 'dist\\share\\icons')

# translations
mkdir('dist\\share\\locale')
locale_dir = os.path.join(gtk_dir, 'share\\locale')
for s in glob.glob('..\\translations\\*.po'):
    lang = s[16:-3]
    # Diffuse localisations
    print 'Compiling %s translation' % (lang, )
    d = 'dist'
    for p in [ 'locale', lang, 'LC_MESSAGES' ]:
        d = os.path.join(d, p)
        mkdir(d)
    d = os.path.join(d, 'diffuse.mo')
    if subprocess.Popen(['msgfmt', '-o', d, s]).wait() != 0:
        raise OSError('Failed to compile "%s" into "%s".' % (s, d))
    # GTK localisations
    d = os.path.join(locale_dir, lang)
    if os.path.isdir(d):
        copyDir(d, os.path.join('dist\\share\\locale', lang))

#
# Add all documentation.
#

# license and other documentation
for p in 'AUTHORS', 'ChangeLog', 'COPYING', 'README':
    copyFile(os.path.join('..', p), os.path.join('dist', p + '.txt'), True)
for p, enc in [ ('ChangeLog_ru', 'cp1251'), ('README_ru', 'cp1251') ]:
    copyFile(os.path.join('..', p), os.path.join('dist', p + '.txt'), True, enc)

# fetch translations for English text hard coded in the stylesheets
translations = {}
f = open('translations.txt', 'rb')
for v in f.read().split('\n'):
    v = v.split(':')
    if len(v) == 3:
        lang = v[0]
        if not translations.has_key(lang):
            translations[lang] = []
        translations[lang].append(v[1:])
f.close()

# convert the manual from DocBook to HTML
d = '..\\src\\usr\\share\\gnome\\help\\diffuse'
for lang in os.listdir(d):
    p = os.path.join(os.path.join(d, lang), 'diffuse.xml')
    if os.path.isfile(p):
        cmd = [ 'xsltproc', '/usr/share/sgml/docbook/xsl-stylesheets/html/docbook.xsl', p ]
        info = subprocess.STARTUPINFO()
        info.dwFlags |= subprocess.STARTF_USESHOWWINDOW
        info.wShowWindow = subprocess.SW_HIDE
        proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, startupinfo=info)
        proc.stdin.close()
        proc.stderr.close()
        fd = proc.stdout
        s = fd.read()
        fd.close()
        if proc.wait() != 0:
            raise OSError('Could not run xsltproc')
        # add link to style sheet
        s = s.replace('</head>', '<link rel="stylesheet" href="style.css" type="text/css"/></head>')
        s = s.replace('<p>\n        </p>', '')
        s = s.replace('<p>\n      </p>', '')
        # cleanup HTML to simpler UTF-8 form
        s = s.replace('<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">', '<meta http-equiv="Content-Type" content="text/html; charset=utf-8">')
        a, idx = [], 0
        while True:
            i = s.find('&#', idx)
            if i < 0:
                a.append(unicode(s[idx:], 'latin_1'))
                break
            a.append(unicode(s[idx:i], 'latin_1'))
            i += 2
            j = s.find(';', i)
            a.append(unichr(int(s[i:j])))
            idx = j + 1
        s = u''.join(a)
        s = codecs.encode(s, 'utf-8')
        # clean up translator credit portion
        div = extract_tag(s, '<div class="othercredit">', '</div>')
        if div is not None:
           firstname = extract_tag(div[2], '<span class="firstname">', '</span>')
           surname = extract_tag(div[2], '<span class="surname">', '</span>')
           contrib = extract_tag(div[2], '<span class="contrib">', '</span>')
           email = extract_tag(div[2], '<code class="email">', '</code>')
           copyright = extract_tag(div[4], '<p class="copyright">', '</p>')
           if firstname is not None and surname is not None and contrib is not None and email is not None and copyright is not None:
               s = '%s%s<p><span class="contrib">%s:</span> <span class="firstname">%s</span> <span class="surname">%s</span> <code class="email">%s</code></p>%s' % (div[0], ''.join(copyright[:4]), contrib[2], firstname[2], surname[2], email[2], copyright[4])
        # translate extra text
        for k, v in translations.get(lang, []):
            s = s.replace(k, v)
        # save HTML version of the manual
        fn = 'manual'
        if lang != 'C':
            fn += '_' + lang
            # update the document language
            s = s.replace(' lang="en" ', ' lang="%s" ' % (lang,))
        f = open(os.path.join('dist', fn + '.html'), 'w')
        f.write(s)
        f.close()
copyFile('style.css', 'dist\\style.css')

#
# Package everything into a single EXE installer.
#

# build binary installer 
copyFile(os.path.join(os.environ['ADD_PATH_HOME'], 'add_path.exe'), 'dist\\add_path.exe')
if os.system('iscc diffuse.iss /F%s' % (INSTALLER, )) != 0:
    raise OSError('Could not run iscc')

#
# Declare success.
#

print 'Successfully created "%s".' % (INSTALLER, )


install.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

# Copyright (C) 2009-2010 Derrick Moser <derrick_moser@yahoo.com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the licence, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program.  You may also obtain a copy of the GNU General Public License
# from the Free Software Foundation by visiting their web site
# (http://www.fsf.org/) or by writing to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import glob
import os
import stat
import subprocess
import sys

app_path = sys.argv[0]

# print a message to stderr
def logError(s):
    sys.stderr.write('%s: %s\n' % (app_path, s))

# this install script should not be used on Windows
if os.name == 'nt':
    logError('Wrong platform.  Use scripts from the "windows-installer" directory instead.')
    sys.exit(1)

# reset the umask so files we create will have the expected permissions
os.umask(stat.S_IWGRP | stat.S_IWOTH)

# option defaults
options = { 'destdir': '/',
            'prefix': '/usr/local/',
            'sysconfdir': '/etc/',
            'examplesdir': '${sysconfdir}',
            'mandir': '${prefix}/share/man/',
            'pythonbin': '/usr/bin/env python' }
install = True
files_only = False

# process --help option
if len(sys.argv) == 2 and sys.argv[1] == '--help':
    print """Usage: %s [OPTION...]

Install or remove Diffuse.

Options:
  --help
     print this help text and quit

  --remove
     remove the program

  --destdir=PATH
     path to the installation's root directory
     default: %s

  --prefix=PATH
     common installation prefix for files
     default: %s

  --sysconfdir=PATH
     directory for installing read-only single-machine data
     default: %s

  --examplesdir=PATH
     directory for example configuration files
     default: %s

  --mandir=PATH
     directory for man pages
     default: %s

  --pythonbin=PATH
     command for python interpreter
     default: %s

  --files-only
     only install/remove files; skip the post install/removal tasks""" % (app_path, options['destdir'], options['prefix'], options['sysconfdir'], options['examplesdir'], options['mandir'], options['pythonbin'])
    sys.exit(0)
 
# returns the list of components used in a path
def components(s):
    return [ p for p in s.split(os.sep) if p != '' ]

# returns a relative path from 'src' to 'dst'
def relpath(src, dst):
    s1, s2, i = components(src), components(dst), 0
    while i < len(s1) and i < len(s2) and s1[i] == s2[i]:
        i += 1
    s = [ os.pardir ] * (len(s1) - i)
    s.extend(s2[i:])
    return os.sep.join(s)

# apply a set of text substitution rules on a string
def replace(s, rules, i=0):
    if i < len(rules):
        k, v = rules[i]
        a = s.split(k)
        for j in range(len(a)):
            a[j] = replace(a[j], rules, i + 1)
        s = v.join(a)
    return s

# create directories
def createDirs(d):
    p = os.sep
    for c in components(d):
        p = os.path.join(p, c)
        if not os.path.isdir(p):
            os.mkdir(p)

# remove a file
def removeFile(f):
    try:
        os.unlink(f)
    except OSError:
        logError('Error removing "%s".' % (f, ))

# install/remove sets of files
def processFiles(install, dst, src, template):
    for k, v in template.items():
        for s in glob.glob(os.path.join(src, k)):
            d = s.replace(src, dst, 1)
            if install:
                createDirs(os.path.dirname(d))
                # install file
                f = open(s, 'rb')
                c = f.read()
                f.close()
                if v is not None:
                    c = replace(c, v)
                print 'Installing %s' % (d, )
                f = open(d, 'wb')
                f.write(c)
                f.close()
                if k == 'bin/diffuse':
                    # turn on the execute bits
                    os.chmod(d, 0755)
            else:
                # remove file
                removeFile(d)

# compile .po files and install
def processTranslations(install, dst):
    for s in glob.glob('translations/*.po'):
        lang = s[13:-3]
        d = os.path.join(dst, 'share/locale/%s/LC_MESSAGES/diffuse.mo' % (lang, ))
        if install:
            # install file
            try:
                print 'Installing %s' % (d, )
                createDirs(os.path.dirname(d))
                if subprocess.Popen(['msgfmt', '-o', d, s]).wait() != 0:
                    raise OSError()
            except OSError:
                logError('WARNING: Failed to compile "%s" localisation.' % (lang, ))
        else:
            # remove file
            removeFile(d)

# parse command line arguments
for arg in sys.argv[1:]:
    if arg == '--remove':
        install = False
    elif arg == '--files-only':
        files_only = True
    else:
        for opt in options.keys():
            key = '--%s=' % (opt, )
            if arg.startswith(key):
                options[opt] = arg[len(key):]
                break
        else:
            logError('Unknown option "%s".' % (arg, ))
            sys.exit(1)

# expand variables
for s in 'sysconfdir', 'examplesdir', 'mandir':
    for k in 'prefix', 'sysconfdir':
        if s != k:
            options[s] = options[s].replace('${%s}' % (k, ), options[k])

# validate inputs
if options['destdir'] == '':
    options['destdir'] = '/'
for opt in 'prefix', 'sysconfdir', 'examplesdir', 'mandir':
    p = options[opt]
    c = components(p)
    if os.pardir in c or os.curdir in c:
        logError('Bad value for option "%s".' % (opt, ))
        sys.exit(1)
    c.insert(0, '')
    c.append('')
    options[opt] = os.sep.join(c)

destdir = options['destdir']
prefix = options['prefix']
sysconfdir = options['sysconfdir']
examplesdir = options['examplesdir']
mandir = options['mandir']
pythonbin = options['pythonbin']

# tell the user what we are about to do
if install:
    stage = 'install'
else:
    stage = 'removal'
print '''Performing %s with:
    destdir=%s
    prefix=%s
    sysconfdir=%s
    examplesdir=%s
    mandir=%s
    pythonbin=%s''' % (stage, destdir, prefix, sysconfdir, examplesdir, mandir, pythonbin)

# install files to prefix
processFiles(install, os.path.join(destdir, prefix[1:]), 'src/usr/', {
        'bin/diffuse': [ ("'../../etc/diffuserc'", repr(relpath(os.path.join(prefix, 'bin'), os.path.join(sysconfdir, 'diffuserc')))), ('/usr/bin/env python', pythonbin) ],
        'share/applications/diffuse.desktop': None,
        'share/diffuse/syntax/*.syntax': None,
        'share/gnome/help/diffuse/*/diffuse.xml': [ ('/usr/', prefix), ('/etc/', sysconfdir) ],
        'share/omf/diffuse/diffuse-*.omf': [ ('/usr/', prefix) ],
        'share/icons/hicolor/*/apps/diffuse.png': None
    })

# install manual
processFiles(install, os.path.join(destdir, mandir[1:]), 'src/usr/share/man/', {
        'man1/diffuse.1': [ ('/usr/', prefix), ('/etc/', sysconfdir) ],
        '*/man1/diffuse.1': [ ('/usr/', prefix), ('/etc/', sysconfdir) ]
    })

# install files to sysconfdir
processFiles(install, os.path.join(destdir, examplesdir[1:]), 'src/etc/', { 'diffuserc': [ ('/etc/', sysconfdir), ('../usr', relpath(sysconfdir, prefix)) ] })

# install translations
processTranslations(install, os.path.join(destdir, prefix[1:]))

if not install:
    # remove directories we own
    for s in 'share/omf/diffuse', 'share/gnome/help/diffuse/C', 'share/gnome/help/diffuse/ru', 'share/gnome/help/diffuse', 'share/diffuse/syntax', 'share/diffuse':
        d = os.path.join(destdir, os.path.join(prefix, s)[1:])
        try:
            os.rmdir(d)
        except OSError:
            logError('Error removing "%s".' % (d, ))

# do post install/removal tasks
if not files_only:
    print 'Performing post %s tasks.' % (stage, )

    cmds = [ [ 'update-desktop-database' ],
             [ 'gtk-update-icon-cache', os.path.join(destdir, os.path.join(prefix, 'icons/hicolor')[1:]) ] ]
    if install:
        cmds.append([ 'scrollkeeper-update', '-q', '-o', os.path.join(destdir, os.path.join(prefix, 'share/omf/diffuse')[1:]) ])
    else:
        cmds.append([ 'scrollkeeper-update', '-q' ])
    for c in cmds:
        for p in os.environ['PATH'].split(os.pathsep):
            if os.path.exists(os.path.join(p, c[0])):
                print ' '.join(c)
                try:
                    if subprocess.Popen(c).wait() != 0:
                        raise OSError()
                except OSError:
                    logError('WARNING: Failed to update documentation database with %s.' % (c[0], ))
                break
        else:
            print 'WARNING: %s is not installed' % (c[0], )
